### PR TITLE
DB-10436 Rollback to Transaction

### PR DESCRIPTION
**Pull request by jpanko1**
New PR after the revert of https://github.com/splicemachine/spliceengine/pull/5103

Jiras following up on discussion in the original PR:

Failure Handling
https://splicemachine.atlassian.net/browse/DB-11394

Handling Running Transactions
https://splicemachine.atlassian.net/browse/DB-11332

Including Flush in Major Compaction
https://splicemachine.atlassian.net/browse/DB-11395

Also, a dependency discovered today that it's not possible to restart Olap Server from java code in K8s:
https://splicemachine.atlassian.net/browse/DBAAS-5061

# Short Description
Allow rollback of entire DB to a past transaction ID.

# Long Description
The original motivation, as noted in the Jira, is that 
> Customer wants to run a whole bunch of script tests - each from the same known starting point over and over again.  So it stands to reason they would restore schema/database to that starting point and rerun

Recommended procedure when running rollback:
1. Connect to DB at snapshot of the transaction ID you want to rollback to.
Example: `connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd;snapshot=7936';`
2. Do sanity check of the DB at this snapshot to be sure it's in the desired state you want to rollback to.
3. Rollback to the transaction ID: `CALL SYSCS_UTIL.SYSCS_ROLLBACK_DATABASE_TO_TRANSACTION();`
4. Connect to current DB.
Example: `connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd';`
5. If running in a K8s environment, restart Olap Server.

Recommend copying and pasting the transaction ID to ensure accuracy.

Note: Requires enterprise access.

# How to test
Make several DB changes.  Get a transaction ID from among the changes.  Rollback to the transaction ID and verify the subsequent changes were rolled back.
Example:

```
create table a ( i int);
insert into a values 1,2;
select * from a;
CALL SYSCS_UTIL.SYSCS_GET_CURRENT_TRANSACTION();  -- 7936
select * from a;
create table b ( i int);
insert into a values 3,4;
select * from a;
select * from b;
CALL SYSCS_UTIL.SYSCS_ENABLE_ENTERPRISE('XXXX');

connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd;snapshot=7936';
-- Verify the DB is in the state you want
CALL SYSCS_UTIL.SYSCS_ROLLBACK_DATABASE_TO_TRANSACTION();
connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd';

select * from a;
select * from b;
```

At the end, `select * from a;` should return 1 and 2.
`select * from b;` should return `ERROR 42X05: Table/View 'B' does not exist.`

More test examples are in the comments of the QA Jira: https://splicemachine.atlassian.net/browse/QA-2773

### CODE REVIEW

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Review by dgomezferro — APPROVED** [review]

File: splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplierImpl.java
Comment (ipraznik-splice) [inline splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplierImpl.java:152]: Commented lines should be removed?

**Review by ipraznik-splice — APPROVED** [review]

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2407/

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2426/

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2432/

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Review by carolp-503 — APPROVED** [review]

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: platform_it/pom.xml
Comment (martinrupp) [inline platform_it/pom.xml:551]: i think you forgot to add the file RollbackToTransactionIT

**Review by martinrupp — CHANGES_REQUESTED** [review]
i think you forgot to add RollbackToTransactionIT
please add a small description what this does and **how to test**.

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: platform_it/pom.xml
Comment (jpanko1) [reply platform_it/pom.xml:551]: The IT is in the EE project because the procedure requires enterprise access.  It's at splice_backup/src/test/java/com/splicemachine/backup/RollbackToTransactionIT.java

I will add the description.

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by martinrupp — COMMENTED** [review]
ah ok thanks

**Review by martinrupp — APPROVED** [review]
approved when edit description

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java
Comment (hatyo) [inline splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java:1071]: this is probably not needed.

File: splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplierImpl.java
Comment (hatyo) [inline splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplierImpl.java:127]: (just copied my comment from [here](https://github.com/splicemachine/spliceengine/pull/5103#discussion_r590202496) to make sure it is not lost :) )

I think we can reduce the complexity of the overlapping algorithm by maintaining the cache online, i.e. whenever we add a new element, we change the cache immediately, provided that we have a suitable equality / hash function and we use a TreeMap, we can probably reduce the complexity to from quadratic to logarithmic. If there is a JIRA open for this, please refer me to it so we can follow up there.

**Review by hatyo — APPROVED** [review]

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplierImpl.java
Comment (jpanko1) [reply splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplierImpl.java:127]: Thanks @hatyo I've added a Jira: https://splicemachine.atlassian.net/browse/DB-11643

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2607/

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2610/

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Review by carolp-503 — APPROVED** [review]
+1 on the new changes once others have approved.

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2617/

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java:716]: Is this needed?

File: splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java:720]: Remove this

**Review by dgomezferro — APPROVED** [review]
Some minor comments

File: splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java
Comment (jpanko1) [reply splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java]: Removed in commit https://github.com/splicemachine/spliceengine/pull/5137/commits/a9487906fa1be8d3a76f234dacd4527eedb05da3

File: splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java
Comment (jpanko1) [reply splice_machine/src/main/java/com/splicemachine/backup/BackupSystemProcedures.java]: Removed printStackTrace in commit https://github.com/splicemachine/spliceengine/pull/5137/commits/a9487906fa1be8d3a76f234dacd4527eedb05da3

**Comment by jpanko1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java
Comment (jpanko1) [reply splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java:1097]: It's followed by rethrowing the exception, so if we're guaranteed that the rethrown exception will be logged or properly handled, the printStackTrace may be redundant.  But keeping it means if the printStackTrace is logged, we know it's logged at least once, and it may be logged in a different log from the one where the rethrown exception will be logged, which could be useful.

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**
